core/drivers/crypto/crypto_api/hash/hash.c

```diff
@@ -0,0 +1,253 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2018-2019 NXP
+ *
+ * Brief   Crypto Hash interface implementation to enable HW driver.
+ */
+
+/* Global includes */
+#include <assert.h>
+#include <crypto/crypto_impl.h>
+#include <malloc.h>
+#include <trace.h>
+#include <utee_defines.h>
+#include <util.h>
+
+/* Driver Crypto includes */
+#include <drvcrypt.h>
+#include <drvcrypt_hash.h>
+
+/*
+ * Format the HASH context to keep the reference to the
+ * operation driver
+ */
+struct crypto_hash {
+	struct crypto_hash_ctx hash_ctx; /* Crypto Hash API context */
+
+	void                 *ctx; /* Hash Context */
+	struct drvcrypt_hash *op;  /* Reference to the operation */
+};
+
+static const struct crypto_hash_ops hash_ops;
+
+/*
+ * Returns the reference to the driver context
+ *
+ * @ctx  API Context
+ *
+ * Returns the reference to the driver context
+ */
+static struct crypto_hash *to_hash_ctx(struct crypto_hash_ctx *ctx)
+{
+	assert(ctx && ctx->ops == &hash_ops);
+
+	return container_of(ctx, struct crypto_hash, hash_ctx);
+}
+
+/*
+ * Checks and returns reference to the driver operations
+ *
+ * @algo  Algorithm
+ * @id    [out] Hash Algorithm internal ID
+ *
+ * Returns the reference to the driver operations
+ */
+static struct drvcrypt_hash *do_check_algo(uint32_t algo,
+						uint8_t *id)
+{
+	struct drvcrypt_hash *hash = NULL;
+	uint8_t algo_op = 0;
+	uint8_t algo_id = 0;
+
+	/* Extract the algorithms fields */
+	algo_op = TEE_ALG_GET_CLASS(algo);
+	algo_id = TEE_ALG_GET_MAIN_ALG(algo);
+
+	if ((algo_op == TEE_OPERATION_DIGEST) &&
+		((algo_id >= TEE_MAIN_ALGO_MD5) &&
+		 (algo_id <= TEE_MAIN_ALGO_SHA512))) {
+
+		hash = drvcrypt_getmod(CRYPTO_HASH);
+
+		/* Verify that the HASH HW implements this algorithm */
+		if (id)
+			*id = algo_id;
+	}
+
+	CRYPTO_TRACE("Check Hash algo %d ret 0x%"PRIxPTR"",
+			algo_id, (uintptr_t)hash);
+
+	return hash;
+}
+
+/*
+ * Free the Software Hashing Context function of the algorithm
+ *
+ * @ctx    Reference the API context
+ */
+static void hash_free_ctx(struct crypto_hash_ctx *ctx)
+{
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+
+	/* Check the parameters */
+	if (hash) {
+		if (hash->op) {
+			if (hash->op->free_ctx)
+				hash->op->free_ctx(hash->ctx);
+		}
+		free(hash);
+	}
+}
+
+/*
+ * Copy Software Hashing Context
+ *
+ * @src_ctx  Reference the API context source
+ * @dst_ctx  [out] Reference the API context destination
+ */
+static void hash_copy_state(struct crypto_hash_ctx *dst_ctx,
+		struct crypto_hash_ctx *src_ctx)
+{
+	struct crypto_hash *hash_src = to_hash_ctx(src_ctx);
+	struct crypto_hash *hash_dst = to_hash_ctx(dst_ctx);
+
+	if ((!dst_ctx) || (!src_ctx))
```

Comment:
should be check before `to_hash_ctx()` is called.

Reply:
In fact you pointed an error, this is the hash_src and hash_dst that must be checked. The to_hash_ctx is already doing the check of dst_ctx and src_ctx.

Reply:
Actually I think `hash_src` as well as `hash_dst` cannot be `NULL`. Maybe `dst_ctx` (and/or `src_ctx`) can be null.
This applies also to `hash_free_ctx()` above and `hash_{init|update|final}()` below.

Note: I one calls `hash_copy_state()` when  `hash_src->op->copy_state == NULL`, this looks like a bug and it should be reposed somehow.

Reply:
Ok, so what do you prefer? 

1. Keep the `to_hash_ctx` as is with the `assert` and remove the check of the `to_hash_ctx` returned?
2. Remove the `assert` and checked if pointer are not NULL?

Then I would say that a driver might not have to implement all operations. The copy_state is fully related to the driver context hence if we don't need to manage a context there is no need for a copy. Again, if you prefer to have all driver functions implemented even if they are empty, I can change this file to panic if a function pointer is NULL

Reply:
I would say option 1.

As for the copy-state function, i think you should do as for [`crypto_hash_copy_state()`](https://github.com/OP-TEE/optee_os/blob/3.6.0/core/crypto/crypto.c#L64).

Reply:
ok